rueidis_test.go

```diff
@@ -197,38 +197,84 @@ func TestFallBackSingleClient(t *testing.T) {
 
 func TestForceSingleClient(t *testing.T) {
 	defer ShouldNotLeaked(SetupLeakDetection())
+
 	ln, err := net.Listen("tcp", "127.0.0.1:0")
 	if err != nil {
 		t.Fatal(err)
 	}
 	defer ln.Close()
-	done := make(chan struct{})
-	go func() {
-		mock, err := accept(t, ln)
-		if err != nil {
-			return
-		}
-		mock.Expect("CLIENT", "SETINFO", "LIB-NAME", LibName).
-			ReplyError("UNKNOWN COMMAND")
-		mock.Expect("CLIENT", "SETINFO", "LIB-VER", LibVer).
-			ReplyError("UNKNOWN COMMAND")
-		mock.Expect("PING").ReplyString("OK")
-		mock.Close()
-		close(done)
-	}()
-	_, port, _ := net.SplitHostPort(ln.Addr().String())
-	client, err := NewClient(ClientOption{
-		InitAddress:       []string{"127.0.0.1:" + port},
-		ForceSingleClient: true,
-	})
-	if err != nil {
-		t.Fatal(err)
-	}
-	if _, ok := client.(*singleClient); !ok {
-		t.Fatal("client should be a singleClient")
+
+	testCases := []struct {
+		name              string
+		maxFailures       int
+		simulateError     bool
+		forceSingleClient bool
+		expectError       bool
+		clientType        string
+	}{
+		{"NoFailuresSingleClient", 0, false, true, false, "*rueidis.singleClient"},
+		{"SimulatedFailuresSingleClient", 3, true, true, true, "*rueidis.singleClient"},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			done := make(chan struct{})
+
+			go func() {
+				mock, err := accept(t, ln)
+				if err != nil {
+					return
+				}
+				mock.Expect("CLIENT", "SETINFO", "LIB-NAME", LibName).
+					ReplyError("UNKNOWN COMMAND")
+				mock.Expect("CLIENT", "SETINFO", "LIB-VER", LibVer).
+					ReplyError("UNKNOWN COMMAND")
+				mock.Expect("PING").ReplyString("OK")
+				mock.Close()
+				close(done)
+			}()
+
+			_, port, _ := net.SplitHostPort(ln.Addr().String())
+
+			dialFn := func(addr string, dialer *net.Dialer, tlsConfig *tls.Config) (net.Conn, error) {
```

Comment:
@rueian Do you think this is how I should do it?
In principle it should work but I am not too sure why it is timing out, maybe I'm missing something

Reply:
The `dailFn` looks good to me but you must invoke `client.Do` to trigger the `dailFn` again after the `NewClient`. Otherwise,  the `dailFn` is only called once.

Reply:
I am slightly stuck here
I should client.Do to trigger the dialFn right after the NewClient call? @rueian

Reply:
Yes. I believe so. There is no other background goroutine that will dial for a connection.